Lizard abilities, wolf heals, and hair powers

Superlagg opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->
[FEATURE][INPUT WANTED]
## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

This gives lizards a set of built-in abilities to change the colors of their skin. They can either swap around their detail colors or change one of them at will, though these abilities cost points to use. 

These points are replenished by eating organs. Some organs, like hearts and brains, are worth more points, while butts are worth nothing and just make you throw up, and cyberorgans deduct points due to being made of metal. This is all handled in a component, listening to a new signal COMSIG_ITEM_CONSUMED, which /obj/item/proc/Eat sends when its done eating a thing.

The component proc is also set to only give lizards ability points, since otherwise lings could freebase a bunch of yee, eat botany, and have infinite DNA points.

This also adds a human-version of the lizard swap-around-your-colors ability as a genetics power. It just lets you swap your hair color around.

This also also shifts that thing where werewolves are healed by eating organs into the component. Works under similar rules as the organ-ability-point thing, except that butts still heal a bit and cyberorgans deal 5 brute and tox. Hearts heal 10 brute and burn, and organs you can't normally eat or find (like heads and chests) are worth 15 of each.

Also adds in some mutant color flags, so cows don't have everything start looking weird if they mess with their hair... horn... things.

Also adds a second entry for customization colors in the appearance holder to hold the player's previous colors when they turn into a mutant. This is so when they go back to being human, their colors don't get messed with by the mutant races clamping their customization colors.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

Lizards are dull, and I want to make them not-dull.

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```
(u)Superlagg:
(*)Lizards can now change up their colors if they eat enough human organs.
(+)Added a mutation that lets people swap their hair colors around.
(+)Increased the healing certain organs give when eaten by a werewolf.
```

Kyle2143 commented:
Without looking at the code, this sounds pretty good to me. It's whacky, has a bit of usefulness, and adds something more unique to lizards. Maybe lizards should get a bit of health from eating human organs similar to the survivalist trait gives you health when eating snacks.

ZeWaka commented:
> Without looking at the code, this sounds pretty good to me. It's whacky, has a bit of usefulness, and adds something more unique to lizards. Maybe lizards should get a bit of health from eating human organs similar to the survivalist trait gives you health when eating snacks. 

Ditto this! Sounds great.

TheMaskedMan2 commented:
This is neat, and to be honest I kind of hope it paves the way for more types of weird little species differences if there is ever more than just Lizards. I always liked playing as weird non-human things.

pali6 commented on code/datums/mutantraces.dm:
```diff
@@ -494,6 +518,15 @@
+	disposing()
+		if(ishuman(mob))
+			var/datum/component/C = mob.GetComponent(/datum/component/consume/organpoints)
+			if (C) C.RemoveComponent(/datum/component/consume/organpoints)
```
```suggestion
			C?.RemoveComponent(/datum/component/consume/organpoints)
```

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+/datum/component/consume
+/datum/component/consume/Initialize()
+	if(!istype(parent, /obj/item))
+		return COMPONENT_INCOMPATIBLE
```
item? really? I thought this was mob-bound

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+	var/list/organ2points = list(/obj/item/organ/head=2,/obj/item/skull=0,/obj/item/organ/brain=3,/obj/item/organ/chest=5,/obj/item/organ/heart=2,/obj/item/organ/appendix=0,/obj/item/clothing/head/butt=0)
+
+/datum/component/consume/organpoints/Initialize()
+	RegisterSignal(parent, list(COMSIG_ITEM_CONSUMED), .proc/eat_organ_get_points)
```
Should check parent type, or rather should likely use the (corrected) `..()` to do that.

pali6 commented on code/datums/mutantraces.dm:
```diff
@@ -494,6 +518,15 @@
+			var/datum/component/C = mob.GetComponent(/datum/component/consume/organpoints)
+			if (C) C.RemoveComponent(/datum/component/consume/organpoints)
+			var/mob/living/carbon/human/L = mob
+			L.abilityHolder.removeAbility(/datum/targetable/lizardAbility/colorshift)
+			L.abilityHolder.removeAbility(/datum/targetable/lizardAbility/colorchange)
```
I wonder if just removing the whole lizard ability holder would work instead. Haven't checked the code but it would likely be cleaner than removing the abilities themselves and leaving the holder dangling there.

Superlagg replied:
I tried doing that, but I couldn't figure put how to make it give the ability holder back if someone stops being a lizard then goes back to being one. Or if they are a lizard and they have their mutantrace set to lizard for whatever reason.

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+
+	if (istype(I) && (I.type in src.organ2hp))
+		M.HealDamage("All", src.organ2hp[I.type] * 0.5, src.organ2hp[I.type] * 0.5)
+		switch(I.type)
```
This excludes synth organs but I think you intended that, just making sure.

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+
+
+/datum/component/consume/organheal
+	var/list/organ2hp = list(/obj/item/organ/head=20,/obj/item/skull=0,/obj/item/organ/brain=30,/obj/item/organ/chest=30,/obj/item/organ/heart=20,/obj/item/organ/appendix=0,/obj/item/clothing/head/butt=6)
```
```suggestion
	var/static/list/organ2hp = list(/obj/item/organ/head=20,/obj/item/skull=0,/obj/item/organ/brain=30,/obj/item/organ/chest=30,/obj/item/organ/heart=20,/obj/item/organ/appendix=0,/obj/item/clothing/head/butt=6)
```
By making it `static` it's not defined anew on each component but only one list exists globally. Saves a bit of time *and* space!

pali6 replied:
Also it'd be much more readable if written like
```
 ... = list(
    /obj/item/organ/head = 20,
    /obj/item/skull = 0,
    ....
  )
```

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+		return 0
+
+	if (istype(I) && (I.type in src.organ2hp))
+		M.HealDamage("All", src.organ2hp[I.type] * 0.5, src.organ2hp[I.type] * 0.5)
+		switch(I.type)
```
Both of these lines exclude subtypes (for example synth heart, various antagonist skulls etc.). Probably intended on your part, just checking.

Superlagg replied:
Was not intended, definitely an oversight. I could have the switch check istype instead of just checking the type.

pali6 replied:
Sadly then it can't be a switch but needs to be either an if-else chain or something else similar.

Superlagg replied:
I used some istyping and a switch! Mostly for flavor text.

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+	var/base_HPup = 5
+
+/datum/component/consume/organheal/Initialize()
+	RegisterSignal(parent, list(COMSIG_ITEM_CONSUMED), .proc/eat_organ_get_heal)
```
Also should probably use (corrected) `..()` to check for compatibility.

Superlagg replied:
Done.

pali6 commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,102 @@
+		return
+
+	if (add_these_points)
+		L.abilityHolder.addPoints(add_these_points, /datum/abilityHolder/lizard)
```
You could make this somewhat more customizable by making the type of the affected ability holder be an argument passed to `Initialize` instead of hardcoding it. I imagine organ eating for points is likely to get reused for something!

Superlagg replied:
Done!

pali6 commented on code/datums/abilities/ability_parent.dm:
```diff
@@ -1191,6 +1197,12 @@
 		for (var/datum/abilityHolder/H in holders)
 			H.deductPoints(cost)
 
+	addPoints(add_points, var/datum/abilityHolder/w_list)
+		for (var/datum/abilityHolder/H in holders)
+			if (w_list && !istype(H, w_list))
+				continue
+			H.addPoints(add_points)
```
```suggestion
	addPoints(add_points, target_ah_type)
		for (var/datum/abilityHolder/H in holders)
			if (w_list && !istype(H, target_ah_type))
				continue
			H.addPoints(add_points)
```
`w_list` is not a very descriptive name, moreover you don't actually want an abilityHolder *object* to get passed here but a *type* so the argument should be untyped.

pali6 commented on code/datums/abilities/ability_parent.dm:
```diff
@@ -223,6 +223,12 @@
 
 		points -= cost
 
+	proc/addPoints(add_points)
```
```suggestion
	proc/addPoints(add_points, target_ah_type)
```
Imo it's good practice to have the same arguments on all overrides of a proc. But opinions might differ.

pali6 commented on _std/macros/appearance.dm:
```diff
@@ -0,0 +1,14 @@
+//non-hairstyle body accessory bitflags
+//corresponds to the color settings in user prefs
+#define DETAIL_1	(1<<0)	//
+#define DETAIL_2	(1<<1)	//
+#define DETAIL_3	(1<<2)	//
```
I feel like these could have some prefix in their name to make them more distinct. If I randomly saw `DETAIL_1` in the code I wouldn't be sure what that means as the name is quite generic.

Superlagg replied:
Done!

Superlagg commented:
Added a new signal to the beginning of Eat() that can be used by components to check if certain inedible things should be edible. Also removed most of the checks on the line that calls Eat(), since those checks were also checked inside the proc, and prevented the signal from doing its job.

This is mostly so lizards and werewolves can eat skulls. Its a bad idea to do so for the most part, as most of them will either stun you or make you puke.

Werewolves can eat severed heads, but lizards can't.

UrsulaMejor commented:
Very cool, but if this goes in we may need to increase the spacebux price of lizard; the ability to change your appearance on the go is about 1/3rd to 1/2 of a traitor item (dna scrambler) in terms of power, and repeatable makes it a very good option for throwing off pursuit.

Tarmunora commented on code/mob.dm:
```diff
@@ -78,6 +78,8 @@
 	var/temp_tolerance = 15 // iterations between each temperature state
 	var/thermoregulation_mult = 0.025 // how quickly the body's temperature tries to correct itself, higher = faster
 	var/innate_temp_resistance = 0.16  // how good the body is at resisting environmental temperature, lower = more resistant
+	var/min_temp_mult = 0.1 // How much should temperature decrease the life mult, if applicable
+	var/max_temp_mult = 5 // How much should temperature increase the life mult, if applicable
```
this seems a very extreme range

Superlagg replied:
Its a debug range, mostly to see if it actually worked. And it does!

Though, I do want a pretty big range. Maybe less than this!

Tarmunora commented on code/mob/living/life/Life.dm:
```diff
@@ -38,6 +38,9 @@
 	proc/get_multiplier()
 		.= clamp(TIME - last_process, tick_spacing, cap_tick_spacing) / tick_spacing
 
+	proc/get_temperature_mult() // mults the mult based on temp
+		. = owner ? clamp((owner.bodytemperature / owner.base_body_temp), owner.min_temp_mult, owner.max_temp_mult) : 1
```
body temperature can go to very high extremes (on both ends of the scale). Perhaps squishifying this in addition to clamping?

Superlagg replied:
I've been trying to figure out a way to do that, which as someone who isnt a math, is kinda tricky!

Tarmunora commented on code/mob/living/life/blood.dm:
```diff
@@ -22,7 +22,8 @@
 
 		var/mult = get_multiplier()
-
+		if(owner && HAS_MOB_PROPERTY(owner, PROP_TEMP_CHEM_EFFECTS))
+			mult *= get_temperature_mult()
```
?? this is blood code, not chems

Superlagg replied:
I should probably name that something better. Or move that function onto the spleen. So, if they're warmer, they should bleed a lot more, but also stop bleeding a lot quicker too.

Tarmunora commented on code/mob/living/life/chems.dm:
```diff
@@ -10,8 +10,12 @@
+
 			owner.reagents.temperature_reagents(owner.bodytemperature, 100, 35/reagent_time_multiplier, 15*reagent_time_multiplier)
 
+			if(owner && HAS_MOB_PROPERTY(owner, PROP_TEMP_CHEM_DEPLETION))
+				reagent_time_multiplier *= get_temperature_mult()
```
this is affecting more than just depletion - this affects the lag-compensation multiplier passed for depletion *and* effects of chems

Superlagg replied:
While I meant for that overall effect, yeah it'd make a lot more sense to have the properties do the things they say they do.

Tarmunora commented on code/mob/living/life/critical.dm:
```diff
@@ -2,6 +2,8 @@
 /datum/lifeprocess/critical //for mobs that use crit (humans only right now)
 	process(var/datum/gas_mixture/environment)
 		var/mult = get_multiplier()
+		if(owner && HAS_MOB_PROPERTY(owner, PROP_TEMP_CRIT))
+			mult *= get_temperature_mult()
```
Hm. I don't really like this - crit multipler is something that I feel should be purely lag-compensation. Seek a second opinion on this perhaps

Tarmunora commented on _std/macros/mob_properties.dm:
```diff
@@ -178,6 +178,10 @@ To remove:
 #define PROP_REBREATHING(x) x("rebreathing", APPLY_MOB_PROPERTY_SIMPLE, REMOVE_MOB_PROPERTY_SIMPLE)
+#define PROP_TEMP_CHEM_DEPLETION(x) x("chem_deplete_heat", APPLY_MOB_PROPERTY_SIMPLE, REMOVE_MOB_PROPERTY_SIMPLE)
+#define PROP_TEMP_CHEM_EFFECTS(x) x("chem_effect_heat", APPLY_MOB_PROPERTY_SIMPLE, REMOVE_MOB_PROPERTY_SIMPLE)
+#define PROP_TEMP_DIGESTION(x) x("digestion_heat", APPLY_MOB_PROPERTY_SIMPLE, REMOVE_MOB_PROPERTY_SIMPLE)
```
doesn't seem to be used?

Superlagg replied:
Not yet, I pushed this in a less-than finished state. For instance, lizards immediately die because most of their organs are missing.

Superlagg commented:
Removed the heat-related crit stuff cus it was rubbish.

Moved the eaten-organ-bonus things to the mutants' intestines, and removed the heat-affected food thing because I couldn't figure out how to get the temperature mult thing to work there without duplicating the code.

Heat-affected livers and kidneys should apply their chem mods to the right places.

Superlagg commented:
Okay I moved the heat-organ stuff to another branch and reverted that stuff. Now, its all just hair powers, wolf heals, and psychedelic lizardpeople!

ZeWaka commented:
is this PR finished and comments addressed? if so, could you resolve the comments?

ZeWaka commented on _std/defines/item.dm:
```diff
@@ -96,4 +96,6 @@
\ No newline at end of file
+/// The pre-attack signal doesnt want the attack to continue, so don't
+#define ATTACK_PRE_DONT_ATTACK 1
+\
```
```suggestion

```

ZeWaka replied:
will review 4real on wednesday afternoon or thursday

ZeWaka commented on _std/defines/item.dm:
```diff
@@ -93,6 +93,8 @@
 #define ITEM_RARITY_MYTHIC 7
 
+// item comp defs
+#define GOOD_4_ME_2_EAT 1
```
better naming pls

Superlagg replied:
Done, now actually named after the thing its supposed to do.

ZeWaka commented on _std/macros/appearance.dm:
```diff
@@ -0,0 +1,14 @@
+#define BODY_DETAIL_OVERSUIT_1		(1<<7)		// Has a detail that goes over the suit, like a cute little enormous cow muzzle
+#define BODY_DETAIL_OVERSUIT_IS_COLORFUL		(1<<8)		// The oversuit is colorful, otherwise don't color it. Defaults to first customization color
+
+#define FIX_COLORS										(1<<9)	// Clamp customization RBG vals between 50 and 190, lizard-style
+#define	HEAD_HAS_OWN_COLORS						(1<<10)	// our head has its own colors that would look weird if tinted
```
```suggestion
/// Clamp customization RBG vals between 50 and 190, lizard-style
#define FIX_COLORS										(1<<9)	
/// our head has its own colors that would look weird if tinted
#define	HEAD_HAS_OWN_COLORS						(1<<10)	
```

ZeWaka commented on _std/macros/appearance.dm:
```diff
@@ -0,0 +1,14 @@
+#define HAS_HAIR_COLORED_HAIR			(1<<3)		// Hair (if any) is/are the color/s you/we (hopefully) set/assigned
+#define HAS_HAIR_COLORED_DETAILS	(1<<5)	// Hair color is used to determine the color of certain non-hair things. Like horns or scales
+
+#define BODY_DETAIL_OVERSUIT_1		(1<<7)		// Has a detail that goes over the suit, like a cute little enormous cow muzzle
+#define BODY_DETAIL_OVERSUIT_IS_COLORFUL		(1<<8)		// The oversuit is colorful, otherwise don't color it. Defaults to first customization color
```
```suggestion
/// Has a detail that goes over the suit, like a cute little enormous cow muzzle
#define BODY_DETAIL_OVERSUIT_1		(1<<7)		
/// The oversuit is colorful, otherwise don't color it. Defaults to first customization color
#define BODY_DETAIL_OVERSUIT_IS_COLORFUL		(1<<8)		
```

ZeWaka commented on _std/macros/appearance.dm:
```diff
@@ -0,0 +1,14 @@
+#define BODY_DETAIL_2	(1<<1)	//
+#define BODY_DETAIL_3	(1<<2)	//
+
+#define HAS_HAIR_COLORED_HAIR			(1<<3)		// Hair (if any) is/are the color/s you/we (hopefully) set/assigned
+#define HAS_HAIR_COLORED_DETAILS	(1<<5)	// Hair color is used to determine the color of certain non-hair things. Like horns or scales
```
```suggestion
/// Hair (if any) is/are the color/s you/we (hopefully) set/assigned
#define HAS_HAIR_COLORED_HAIR			(1<<3)		
/// Hair color is used to determine the color of certain non-hair things. Like horns or scales
#define HAS_HAIR_COLORED_DETAILS	(1<<5)	
```
please use the proper documentation style so it gets picked up

ZeWaka commented on code/datums/abilities/lizard.dm:
```diff
@@ -0,0 +1,155 @@
+			L = holder.owner
+		return
+
+	proc/fix_colors(var/hex)
```
describe the function of this proc... i might know what it does but others won't

Superlagg replied:
Thought I already described them, but I guess that was another pr. Added a comment.

ZeWaka commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,238 @@
+/datum/component/consume
```
can you describe the purpose of this component a bit

Superlagg replied:
Done, added a comment that explains the thing it does.

ZeWaka commented on code/datums/components/food_stuff.dm:
```diff
@@ -0,0 +1,238 @@
+				if (/obj/item/organ/heart/synth)
+					boutput(L, "<span class='notice'>Full of pharosium!</span>")
+				if (/obj/item/organ/heart/flock)
+					boutput(L, "<span class='notice'>Full of iron...?</span>")
```
i would say gnesis or blue goop, but vOv

Superlagg replied:
Changed to taste like adjective chicken.

ZeWaka commented on code/datums/mutantraces.dm:
```diff
@@ -314,6 +313,43 @@
 		..()
 		return
 
+	proc/fix_colors(var/hex)
```
ditto from before

ZeWaka commented on code/datums/mutantraces.dm:
```diff
@@ -464,38 +500,23 @@
 		..()
-		if(ishuman(mob))
-			var/datum/appearanceHolder/aH = mob.bioHolder.mobAppearance
+		if(ishuman(H))
```
```suggestion
		if(istype(H))
```

ZeWaka replied:
if you really want to check, but this should always be passed a human....

ZeWaka commented on code/mob/living/carbon/human/procs/update_icon.dm:
```diff
@@ -1190,11 +1192,30 @@ var/list/update_body_limbs = list("r_arm" = "stump_arm_right", "l_arm" = "stump_
+				src.body_standing.overlays += mutant_image
+
+			if (AHol.mob_color_flags & BODY_DETAIL_OVERSUIT_1)
+				mutant_image = src.mutantrace.detail_over_suit
```
```suggestion
				mutant_image = src.mutantrace.detail_over_suit // used by lizards so that when they're facing upwards their tail doesn't look awkward.
```

ZeWaka commented on code/modules/medical/genetics/bioEffects/powers.dm:
```diff
@@ -485,6 +485,65 @@
 ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
```suggestion
/*  / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / /  */
/* / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / /  */
```

ZeWaka commented on code/modules/medical/genetics/bioEffects/powers.dm:
```diff
@@ -485,6 +485,65 @@
+		if (istype(H.mutantrace, /datum/mutantrace/lizard))
+			boutput(H, "<span class='notice'>You don't have any hair!</span>")
+			return
+		else if (H?.mutantrace?.override_hair && !istype(H.mutantrace, /datum/mutantrace/cow))
```
```suggestion
		else if (H.mutantrace?.override_hair && !istype(H.mutantrace, /datum/mutantrace/cow))
```
guarenteed

ZeWaka commented on code/modules/medical/genetics/bioEffects/powers.dm:
```diff
@@ -485,6 +485,65 @@
+			boutput(H, "<span class='notice'>Whatever hair you have isn't affected!</span>")
+			return
+
+		if (H?.bioHolder?.mobAppearance)
```
```suggestion
		if (H.bioHolder?.mobAppearance)
```
guarenteed

ZeWaka commented on code/modules/medical/genetics/bioEffects/powers.dm:
```diff
@@ -485,6 +485,65 @@
+			H.update_body()
+			H.update_face()
+
+		else // ???
+			return
```
dead code
```suggestion
```

ZeWaka commented on code/modules/medical/genetics/bioEffects/powers.dm:
```diff
@@ -485,6 +485,65 @@
+
+		else // ???
+			return
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
```suggestion
/* / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / */
/* / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / */
/* / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / / */
```

ZeWaka commented on code/modules/medical/genetics/bioHolder.dm:
```diff
@@ -16,6 +16,10 @@ var/list/datum/bioEffect/mutini_effects = list()
 
 /datum/appearanceHolder
 	//Holds all the appearance information.
```
```suggestion
/// Holds all the appearance information.
/datum/appearanceHolder	
```

ZeWaka commented on goonstation.dme:
```diff
@@ -25,6 +25,7 @@ var/datum/preMapLoad/preMapLoad = new
 #include "code\__secret_public.dme"
 
 // BEGIN_INCLUDE
+#include "_std\macros\appearance.dm"
```
no.

Superlagg replied:
No?

ZeWaka replied:
This is not how _std folder includes are done

Superlagg replied:
I probably should've been paying attention when Adhara was cleaning all this stuff up. Moved to the right file.

Kyle2143 reviewed (approved):
I think it's good to add, but make zewaka's changes first. and then it's good to go.
